### PR TITLE
NAVAND-1148: introduce RoutesRenderedCallback

### PR DESCRIPTION
**Pull request by dzinad**

### CODE REVIEW

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ Callback to be notified of rendered routes.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+interface RoutesRenderedCallback {
+
+    /**
+     * Map instance where the routes are expected to be rendered.
+     */
+    val map: MapboxMap
```
Comment (VysotskiVadim) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:16]: Maybe it should be a parameter rather than a part of the callback?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * @param successfullyRenderedRouteIds ids of the routes that were successfully rendered (correspond to [NavigationRoute.id])
+ * @param renderingCancelledRouteIds ids of the routes whose rendering was cancelled (correspond to [NavigationRoute.id]).
+ *  This may happen, for example, when a newer rendering operation had been queued before the previous one finished.
+ * @param successfullyClearedRouteIds ids of the routes that were successfully cleared (correspond to [NavigationRoute.id])
+ * @param clearingCancelledRouteIds ids of the routes whose clearance was cancelled (correspond to [NavigationRoute.id]).
+ *  This may happen, for example, when a newer rendering operation had been queued before the previous one finished.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class RoutesRenderedResult internal constructor(
+    val successfullyRenderedRouteIds: Set<String>,
```
Comment (VysotskiVadim) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:38]: We call the callback for a group of routes, while under the hood they are drown one by one. It can cause the opposite effect: route objects won't be drown until all 3 routes are visible.

**Review by VysotskiVadim — COMMENTED** [review]
I think that current implementation is good enough for the first iteration. We have freedom to change the API in future as it's marked as experimental. And I think this feature deserves one more iteration.
I also find the implementation complex, however I understand that this complexity is unavoidable with current approach. I don't see any obvious errors base on what I've learned about route line so far, well done @dzinad 💪  , but complex things tend to break more often. On the second iteration I would think about decreasing complexity.

I think, we need to look into direction of `renderedRoutesObserver` which is triggered every time rendered routes changes, so that:
1. User can draw route objects one by one as soon as each route is rendered
2. We don't need to batch updates from maps on the SDK side, just proxy callbacks as rendered routes changes

I hope, that observer instead of callback can make thing simpler for both SDK and apps.

@LukasPaczos , I hope you will be able to look through as well. I don't feel to have enough context and experience with route line to approve this change. In case you aren't able to review, please let me know, I will dig deeper to be able to approve it 🙂

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt [hunk 3]
```diff
@@ -452,33 +542,63 @@ class MapboxRouteLineView(options: MapboxRouteLineOptions) {
...
+                                    routeId
+                                )
+                                sourceToFeatureMap[routeLineSourceKey] = RouteLineFeatureId(routeId)
                             }
                         }
+                    if (callback != null) {
+                        routesExpector.expectRoutes(
+                            expectedRoutesData.allRenderedRouteIds,
+                            expectedRoutesData.allClearedRouteIds,
+                            expectedRoutesData,
```
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt:597]: We're already passing the whole `expectedRoutesData`, why do we need to pass IDs separately?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/DataIdHolder.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.mapbox.navigation.ui.maps.route.line.api
+
+internal class DataIdHolder {
+
+    private val sourceIdToDataIdAndRouteId = mutableMapOf<String, Int>()
+    fun incrementDataId(sourceId: String): Int {
+        val newDataId = (sourceIdToDataIdAndRouteId[sourceId] ?: 0) + 1
+        sourceIdToDataIdAndRouteId[sourceId] = newDataId
+        return newDataId
+    }
```
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/DataIdHolder.kt:10]: Oh yes, it matters a lot.
TLDR: dataId is an **integer** field that we receive in `onEventDataLoaded` callback if we passed it previously when invoking `featureCollection`. It means that this callback corresponds to this operation. It also means that all previous operations for this source id either already finished or are cancelled. We increment dataId for each operation on this particular sourceId so that we can later keep track of what operations we can consider cancelled.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    val renderingCancelledRouteIds: Set<String>,
+    val successfullyClearedRouteIds: Set<String>,
+    val clearingCancelledRouteIds: Set<String>,
```
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:41]: renderedRoute means that the route is drawn on map, cleared - that it's cleared.
1. We can invoke `rendreRouteDrawData` with empty feature collection - routes will be erased.
2. We can invoke `clearRouteLineValue` with non-empty routes - they will be drawn.
So it's possible that we receive non-empty cleared routes for `renderRouteDrawData` and non-empty rendered routes for `clearRouteLineValue`.

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/RouteLineLayersTest.kt [hunk 6]
```diff
@@ -587,6 +592,321 @@ class RouteLineLayersTest : BaseTest<BasicNavigationViewActivity>(
...
+        )
+        assertEquals(
+            RoutesRenderedResultFactory.routesRenderedResult(
+                setOf(route1.id, route2.id, route4.id),
+                emptySet(),
+                setOf(route3.id),
+                emptySet(),
+            ),
+            renderedRoutesResult10
+        )
```
[the remark this answers is not in the document]
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/RouteLineLayersTest.kt:840]: Not really. 
Both this fields are the result of the operation in which the callback was used.
Not actually redrawing the routes on map doesn't mean they weren't rendered. They were. As a result of this operation. The mechanism under-the-hood was different. Consider this an optimization: we could as well clear the route line and rerender it, but the callback logic shouldn't depend on these details.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:41]: We still need to keep track of these ids to understand when all routes are processed. And after that we can either expose that info or not. IMO for consistency we should expose it (so that the passed ids are never "lost").

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 1 above)
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:16]: It can, I just thought that it always comes with a callback, so why not store it there? I don't have a strong opinion here.
@LukasPaczos any thoughts?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 2 above)
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:38]: Yes, but the alternative is to have callback be invoked multiple number of times, which doesn't fit into callback pattern.
The observer pattern was discarded earlier, more context: https://github.com/mapbox/mapbox-navigation-android/pull/6964#issuecomment-1441621153
And a technical limitation: https://github.com/mapbox/mapbox-navigation-android/pull/6964#discussion_r1116185159

File: changelog/unreleased/features/7058.md [hunk 7]
```diff
@@ -0,0 +1 @@
+- Introduced `RoutesRenderedCallback` along with overrides to `MapboxRouteLineView#renderRouteDrawData` and `MapboxRouteLineView#renderClearRouteLineValue` to notified whenever the routes are rendered on map / cleared from map.
```
[the remark this answers is not in the document]
Comment (dzinad) [reply changelog/unreleased/features/7058.md:1]: Right...

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/DataIdHolder.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/DataIdHolder.kt:10]: BTW, I'm also sorry that you have to look at this code. :)
It's too complicated for my taste, but I couldn't find a better way. It took me a couple of iterations to a) make the logic consistent; b) simplify the code a bit :) to achieve this.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:16]: There's a small chance of a memory leak if someone holds onto the callback object beyond the lifecycle of the map. It's an edge-case but it's still a risk we introduce with this type of an API. If we don't have a strong opinion, then passing a map as a `MapboxRouteLineView` parameter might be better because then we can make `RoutesRenderedCallback` a SAM interface.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt:41]: > We can invoke `clearRouteLineValue` with non-empty routes

Ah, right, I forgot about this fact. My thoughts on this are in https://github.com/mapbox/mapbox-navigation-android/pull/5859#discussion_r896710181.

**Review by LukasPaczos — APPROVED** [review]
LGTM (only with https://github.com/mapbox/mapbox-navigation-android/pull/7058#discussion_r1154364721 to consider). The code is certainly complex and there is room for us to revisit and simplify (with changes to other components of route rendering or even Maps SDK) but great job navigating that complexity @dzinad and getting us to a working solution.

**Comment by LukasPaczos** [comment]
Side note - small, internal documentation committed somewhere in code on why this setup works as it does would go a long way.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt (hunk 1 above)
Comment (dzinad) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/RoutesRenderedCallback.kt]: Addressed.

**Comment by dzinad** [comment]
> Side note - small, internal documentation committed somewhere in code on why this setup works as it does would go a long way.

Will add. Maybe in another PR if builds are green here.

**Comment by dzinad** [comment]
@LukasPaczos @VysotskiVadim added docs: https://github.com/mapbox/mapbox-navigation-android/pull/7071.
